File: crates/cxx-qt-lib/src/core/qtimezone.cpp
// clang-format off
// SPDX-FileCopyrightText: 2023 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// clang-format on
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0
#include "cxx-qt-lib/qtimezone.h"

namespace rust {
namespace cxxqtlib1 {

QList<QByteArray>
qtimezoneAvailableTimeZoneIds()
{
  return QTimeZone::availableTimeZoneIds();
}

::std::unique_ptr<QTimeZone>
qtimezoneClone(const QTimeZone& timezone)
{
  return ::std::make_unique<QTimeZone>(timezone);
}

::std::unique_ptr<QTimeZone>
qtimezoneDefault()
{
  return ::std::make_unique<QTimeZone>();
}

QString
qtimezoneDisplayName(const QTimeZone& timezone,
                     QTimeZoneTimeType timeType,
                     QTimeZoneNameType nameType)
{
  return timezone.displayName(timeType, nameType);
}

::std::unique_ptr<QTimeZone>
qtimezoneFromOffsetSeconds(::std::int32_t offsetSeconds)
{
  return ::std::make_unique<QTimeZone>(static_cast<int>(offsetSeconds));
}

::std::unique_ptr<QTimeZone>
qtimezoneFromIana(const QByteArray& ianaId)
{
  return ::std::make_unique<QTimeZone>(ianaId);
}

::std::unique_ptr<QTimeZone>
qtimezoneSystemTimeZone()
{
  return ::std::make_unique<QTimeZone>(QTimeZone::systemTimeZone());
}

QByteArray
qtimezoneSystemTimeZoneId()
{
  return QTimeZone::systemTimeZoneId();
}

::std::unique_ptr<QTimeZone>
qtimezoneUtc()
{
  return ::std::make_unique<QTimeZone>(QTimeZone::utc());
}

}
}

File: tests/qt_types_standalone/rust/src/qtimezone.rs
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0

use cxx_qt_lib::{QByteArray, QTimeZone};

#[cxx::bridge]
mod qtimezone_cxx {
    unsafe extern "C++" {
        include!("cxx-qt-lib/qtimezone.h");
        type QTimeZone = cxx_qt_lib::QTimeZone;
    }

    extern "Rust" {
        fn construct_qtimezone() -> QTimeZone;
        fn read_qtimezone(t: &QTimeZone) -> bool;
        fn clone_qtimezone(t: &QTimeZone) -> QTimeZone;
    }
}

fn construct_qtimezone() -> QTimeZone {
    QTimeZone::from_iana(&QByteArray::from("Europe/London"))
}

fn read_qtimezone(t: &QTimeZone) -> bool {
    t.id().to_string() == "Europe/London"
}

fn clone_qtimezone(t: &QTimeZone) -> QTimeZone {
    t.clone()
}
